msal/imds.py

```diff
@@ -0,0 +1,164 @@
+# Copyright (c) Microsoft Corporation.
+# All rights reserved.
+#
+# This code is licensed under the MIT License.
+import json
+import logging
+import os
+import socket
+import time
+try:  # Python 2
+    from urlparse import urlparse
+except:  # Python 3
+    from urllib.parse import urlparse
+
+logger = logging.getLogger(__name__)
+
+def _scope_to_resource(scope):  # This is an experimental reasonable-effort approach
+    u = urlparse(scope)
+    if u.scheme:
+        return "{}://{}".format(u.scheme, u.netloc)
+    return scope  # There is no much else we can do here
```

Comment:
Azure CLI's implementation only removes `/.default`, `/user_impersonation` suffix.

https://github.com/Azure/azure-cli/blob/8474a10fc16b5a168f96dbbe0fd94afc8eecdf4e/src/azure-cli-core/azure/cli/core/auth/util.py#L77-L96

Reply:
This experimental helper is currently NOT being used and not really exposed. And we may remove it later.

Do we have some documentation that defines a score-to-scope algorithm? CC: @jmprieur , @bgavrilMS

Reply:
> Do we have some documentation that defines a canonical score-to-scope algorithm?

Well, I don't think so. `scope` is like the version 2.0 of `resource`. Converting `scope` back to `resource` certainly losses some information.

Reply:
@rayluo - consider supportting both? The public API can mention resource, but if the user passes in the scope (i.e. graph/.default, just strip the suffix)

Reply:
> @jiasli - for Managed Identity / S2S tokens, there is a 1:1 mapping between scope and resource.
> 
> E.g. **resource**: https://graph.microsoft.com **scope**: https://graph.microsoft.com/.default

Can the `https://resource/user_impersonation` mentioned in @jiasli's first message above also be converted to `https://resource`? (You see, I am really open to add this `scope_to_resource()` helper. I just want to have a precise algorithm to follow.)

Reply:
I gave it more thoughts.

In the absence of a written documentation on "what suffixes can be safely trimmed in a scope-to-resource conversion for AAD", let's follow this [conceptual doc "Scopes and permissions in the Microsoft identity platform"](https://learn.microsoft.com/en-us/azure/active-directory/develop/scopes-oidc) and the empirical evidence in [this video (although it was mainly talking about the .default permission)](https://dev.to/425show/just-what-is-the-default-scope-in-the-microsoft-identity-platform-azure-ad-2o4d), that scopes are typically in the form of `https://resource_uri/permission`. In that sense, the algorithm in this PR's helper is functional.

The `ManagedIdentity.acquire_token(resource)` will remain accepting resource only. Together with the `scope_to_resource()` helper at the developer's disposal, we are technically supporting both. But `acquire_token(resource)` does not automatically call `scope_to_resource()` conversion, because the latter would not know whether the developer would want the trailing `/` (currently, the `scope_to_resource()` helper removes it). It is a responsibility of downstream caller - and ultimately the human developer/user - to decide what resource should be used to satisfy the token audience.

CC: @bgavrilMS